diff --git a/autocomplete_deluxe.module b/autocomplete_deluxe.module
index d69de17..88ae88e 100755
--- a/autocomplete_deluxe.module
+++ b/autocomplete_deluxe.module
@@ -351,11 +351,14 @@ function autocomplete_deluxe_after_build($element, &$form_state) {
     $element['#value'] = trim($element['#value']);
     // Replace all double double quotes and space with a comma. This will allows
     // us to keep entries in double quotes.
-    $element['#value'] = str_replace('"" ""', ',', $element['#value']);
-    $element['#value'] = str_replace('""  ""', ',', $element['#value']);
+    $element['#value'] = str_replace('"" ""', '", "', $element['#value']);
+    $element['#value'] = str_replace('""  ""', '", "', $element['#value']);
     // Remove the double quotes at the beginning and the end from the first and
     // the last term.
-    $element['#value'] = substr($element['#value'], 2, strlen($element['#value']) - 4);
+    $element['#value'] = substr($element['#value'], 1, strlen($element['#value']) - 2);
+    // One final check to turn all double double quotes into single double quotes. 
+    // Necessary when re-saving existing values.
+    $element['#value'] = str_replace('""', '"', $element['#value']);
 
     unset($element['value_field']['#maxlength']);
   }
